File: apps/chat/src/collectors.ts
import { randomUUID } from "node:crypto";

import type { ChatProvider, ChatSourceId, ChatStreamEvent } from "@coldbrew/packages/chat.js";
import { logger } from "@coldbrew/packages/logger.js";
import { delay } from "@lebedevna/delay";

import type { ChatEventBroker } from "./chat-application.js";
import type { CollectorLease } from "./nats.js";
import type { ChatProviderAdapter, ConnectedChatSource } from "./provider.js";

const RECONCILE_INTERVAL_MS = 10_000;
const LEASE_RETRY_MS = 5_000;
const STREAM_RETRY_MS = 2_000;

type OwnedConnectedSource = Readonly<{
  userId: number;
  connectedSource: ConnectedChatSource;
}>;

export interface ChatCollectorStore {
  getAllEnabledSources(): Promise<readonly OwnedConnectedSource[]>;
}

export interface ChatCollectorRefreshControl {
  refreshes(signal: AbortSignal): AsyncIterable<ChatSourceId>;
}

export interface ChatCollectorLeases {
  acquire(key: string, owner: string): Promise<CollectorLease | null>;
}

function eventKey(event: ChatStreamEvent) {
  if (event.type === "message") return `message:${event.message.sourceId}:${event.message.id}`;
  if (event.type === "message_deleted") {
    return `deleted:${event.sourceId}:${event.messageId}`;
  }
  return `${event.type}:${event.type === "state" ? event.sourceId : "connection"}:${randomUUID()}`;
}

async function collectSource(
  ownedSource: OwnedConnectedSource,
  provider: ChatProviderAdapter,
  broker: ChatEventBroker,
  leases: ChatCollectorLeases,
  serviceOwner: string,
  signal: AbortSignal,
) {
  const { userId, connectedSource } = ownedSource;
  const { sourceId } = connectedSource.source;

  while (!signal.aborted) {
    const lease = await leases.acquire(sourceId, serviceOwner);
    if (!lease) {
      await delay(LEASE_RETRY_MS, { signal });
      continue;
    }

    const leaseController = new AbortController();
    const collectionSignal = AbortSignal.any([signal, leaseController.signal]);
    const maintaining = lease.maintain(collectionSignal).catch((cause: unknown) => {
      if (!collectionSignal.aborted) logger.error({ cause, sourceId }, "Chat collector lease lost");
      leaseController.abort();
    });

    try {
      for await (const $event of provider.stream(connectedSource, collectionSignal)) {
        if ($event.isErr()) {
          await broker.publish(
            userId,
            {
              type: "state",
              sourceId,
              state: "error",
              detail: $event.error.detail,
            },
            `error:${sourceId}:${randomUUID()}`,
          );
          continue;
        }
        await broker.publish(userId, $event.value, eventKey($event.value));
      }
    } catch (cause) {
      if (!collectionSignal.aborted) logger.error({ cause, sourceId }, "Chat collector failed");
    } finally {
      leaseController.abort();
      await maintaining;
      await lease.release().catch((cause: unknown) => {
        if (!signal.aborted)
          logger.error({ cause, sourceId }, "Chat collector lease release failed");
      });
    }

    if (!signal.aborted) await delay(STREAM_RETRY_MS, { signal });
  }
}

export async function runChatCollectors(
  store: ChatCollectorStore,
  broker: ChatEventBroker,
  leases: ChatCollectorLeases,
  collectorControl: ChatCollectorRefreshControl,
  providers: readonly ChatProviderAdapter[],
  signal: AbortSignal,
) {
  const serviceOwner = randomUUID();
  const providerByName = new Map<ChatProvider, ChatProviderAdapter>(
    providers.map((provider) => [provider.provider, provider]),
  );
  const running = new Map<
    string,
    Readonly<{ abort: () => void; generation: symbol; work: Promise<void>; tokenVersion: number }>
  >();
  let reconciliation = Promise.resolve();

  const reconcile = async () => {
    const enabledSources = await store.getAllEnabledSources();
    const desired = new Map(
      enabledSources
        .filter(({ connectedSource }) => {
          return providerByName.get(connectedSource.source.provider)?.collection === "pull";
        })
        .map(({ connectedSource }) => [connectedSource.source.sourceId, connectedSource] as const),
    );

    for (const [sourceId, collector] of running) {
      const desiredSource = desired.get(sourceId);
      if (desiredSource?.credentials.tokenVersion === collector.tokenVersion) continue;
      collector.abort();
      running.delete(sourceId);
    }

    for (const ownedSource of enabledSources) {
      const source = ownedSource.connectedSource.source;
      const provider = providerByName.get(source.provider);
      if (!provider || provider.collection !== "pull" || running.has(source.sourceId)) continue;

      const controller = new AbortController();
      const generation = Symbol(source.sourceId);
      const collectorSignal = AbortSignal.any([signal, controller.signal]);
      const work = collectSource(
        ownedSource,
        provider,
        broker,
        leases,
        serviceOwner,
        collectorSignal,
      ).finally(() => {
        if (running.get(source.sourceId)?.generation === generation) {
          running.delete(source.sourceId);
        }
      });
      running.set(source.sourceId, {
        abort: () => controller.abort(),
        generation,
        work,
        tokenVersion: ownedSource.connectedSource.credentials.tokenVersion,
      });
    }
  };

  const serialized = (work: () => Promise<void>) => {
    const next = reconciliation.then(work);
    reconciliation = next.catch((cause: unknown) =>
      logger.error({ cause }, "Chat collector reconciliation failed"),
    );
    return next;
  };

  const refreshing = (async () => {
    for await (const sourceId of collectorControl.refreshes(signal)) {
      await serialized(async () => {
        const collector = running.get(sourceId);
        if (!collector) return;
        collector.abort();
        running.delete(sourceId);
        await collector.work;
        if (!signal.aborted) await reconcile();
      });
    }
  })();

  try {
    while (!signal.aborted) {
      await serialized(reconcile);
      await delay(RECONCILE_INTERVAL_MS, { signal });
    }
  } finally {
    for (const collector of running.values()) collector.abort();
    await Promise.allSettled([...running.values()].map(({ work }) => work));
    await refreshing;
  }
}

File: internal/chat/collectors.go
package chat

import (
	"context"
	"fmt"
	"log/slog"
	"time"
)

const (
	collectorReconcileInterval = 10 * time.Second
	collectorLeaseRetry        = 5 * time.Second
	collectorStreamRetry       = 2 * time.Second
)

type CollectorStore interface {
	GetAllEnabledSources(context.Context) ([]OwnedConnectedSource, error)
}

type CollectorLeases interface {
	Acquire(context.Context, string, string) (CollectorLease, error)
}

type CollectorRefreshControl interface {
	Refreshes(context.Context) (<-chan string, error)
}

type runningCollector struct {
	cancel       context.CancelFunc
	done         <-chan struct{}
	tokenVersion int
}

func RunCollectors(ctx context.Context, store CollectorStore, broker EventBroker, leases CollectorLeases, control CollectorRefreshControl, providers []Provider) error {
	providerByName := make(map[string]Provider, len(providers))
	for _, provider := range providers {
		providerByName[provider.Name()] = provider
	}
	refreshes, err := control.Refreshes(ctx)
	if err != nil {
		return err
	}
	serviceOwner := randomID()
	running := make(map[string]runningCollector)
	reconcile := func() error {
		sources, err := store.GetAllEnabledSources(ctx)
		if err != nil {
			return err
		}
		desired := make(map[string]OwnedConnectedSource)
		for _, source := range sources {
			provider := providerByName[source.ConnectedSource.Source.Provider]
			if provider != nil && provider.Collection() == "pull" {
				desired[source.ConnectedSource.Source.SourceID] = source
			}
		}
		for sourceID, collector := range running {
			source, exists := desired[sourceID]
			if exists && source.ConnectedSource.Credentials.TokenVersion == collector.tokenVersion {
				continue
			}
			collector.cancel()
			<-collector.done
			delete(running, sourceID)
		}
		for sourceID, source := range desired {
			if _, exists := running[sourceID]; exists {
				continue
			}
			collectorCtx, cancel := context.WithCancel(ctx)
			done := make(chan struct{})
			provider := providerByName[source.ConnectedSource.Source.Provider]
			go func() {
				defer close(done)
				collectSource(collectorCtx, source, provider, broker, leases, serviceOwner)
			}()
			running[sourceID] = runningCollector{cancel: cancel, done: done, tokenVersion: source.ConnectedSource.Credentials.TokenVersion}
		}
		return nil
	}
	if err := reconcile(); err != nil {
		return err
	}
	ticker := time.NewTicker(collectorReconcileInterval)
	defer ticker.Stop()
	defer func() {
		for _, collector := range running {
			collector.cancel()
		}
		for _, collector := range running {
			<-collector.done
		}
	}()
	for {
		select {
		case <-ctx.Done():
			return nil
		case sourceID, open := <-refreshes:
			if !open {
				refreshes = nil
				continue
			}
			if collector, exists := running[sourceID]; exists {
				collector.cancel()
				<-collector.done
				delete(running, sourceID)
			}
			if err := reconcile(); err != nil {
				slog.Error("Chat collector reconciliation failed", "error", err)
			}
		case <-ticker.C:
			if err := reconcile(); err != nil {
				slog.Error("Chat collector reconciliation failed", "error", err)
			}
		}
	}
}

func collectSource(ctx context.Context, owned OwnedConnectedSource, provider Provider, broker EventBroker, leases CollectorLeases, owner string) {
	sourceID := owned.ConnectedSource.Source.SourceID
	for ctx.Err() == nil {
		lease, err := leases.Acquire(ctx, sourceID, owner)
		if err != nil {
			slog.Error("Chat collector lease acquisition failed", "sourceId", sourceID, "error", err)
			if !waitFor(ctx, collectorLeaseRetry) {
				return
			}
			continue
		}
		if lease == nil {
			if !waitFor(ctx, collectorLeaseRetry) {
				return
			}
			continue
		}
		collectionCtx, cancel := context.WithCancel(ctx)
		maintained := make(chan error, 1)
		go func() { maintained <- lease.Maintain(collectionCtx) }()
		events, providerErrors := provider.Stream(collectionCtx, owned.ConnectedSource)
		collecting := true
		maintenanceFinished := false
		for collecting && collectionCtx.Err() == nil {
			select {
			case <-collectionCtx.Done():
				collecting = false
			case err := <-maintained:
				maintenanceFinished = true
				if err != nil {
					slog.Error("Chat collector lease lost", "sourceId", sourceID, "error", err)
				}
				collecting = false
			case event, open := <-events:
				if !open {
					events = nil
					if providerErrors == nil {
						collecting = false
					}
					continue
				}
				if err := broker.Publish(collectionCtx, owned.UserID, event, eventKey(event)); err != nil {
					slog.Error("publish chat event", "sourceId", sourceID, "error", err)
					collecting = false
				}
			case providerErr, open := <-providerErrors:
				if !open {
					providerErrors = nil
					if events == nil {
						collecting = false
					}
					continue
				}
				if providerErr != nil {
					event := StreamEvent{Type: "state", SourceID: sourceID, State: "error", Detail: providerErrorDetail(providerErr)}
					if err := broker.Publish(collectionCtx, owned.UserID, event, eventKey(event)); err != nil {
						slog.Error("publish chat provider error", "sourceId", sourceID, "error", err)
						collecting = false
					}
				}
			}
		}
		cancel()
		if !maintenanceFinished {
			<-maintained
		}
		if err := lease.Release(); err != nil && ctx.Err() == nil {
			slog.Error("Chat collector lease release failed", "sourceId", sourceID, "error", err)
		}
		if !waitFor(ctx, collectorStreamRetry) {
			return
		}
	}
}

func eventKey(event StreamEvent) string {
	if event.Type == "message" && event.Message != nil {
		return "message:" + event.Message.SourceID + ":" + event.Message.ID
	}
	if event.Type == "message_deleted" {
		return "deleted:" + event.SourceID + ":" + event.MessageID
	}
	source := event.SourceID
	if source == "" {
		source = "connection"
	}
	return fmt.Sprintf("%s:%s:%s", event.Type, source, randomID())
}

func waitFor(ctx context.Context, duration time.Duration) bool {
	timer := time.NewTimer(duration)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return false
	case <-timer.C:
		return true
	}
}

var _ CollectorStore = (*Store)(nil)
var _ CollectorLeases = (*NatsCollectorLeases)(nil)
var _ CollectorRefreshControl = (*NatsCollectorControl)(nil)
